src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/NuGetProjectService.cs

```diff
@@ -0,0 +1,104 @@
+        }
+
+        public async Task<InstalledPackagesResult> GetInstalledPackagesAsync(Guid projectId, CancellationToken cancellationToken)
+        {
+            // Just in case we're on the UI thread, switch to background thread. Very low cost (does not schedule new task) if already on background thread.
+            await TaskScheduler.Default;
+
+            NuGetProject project = await _solutionManager.GetNuGetProjectAsync(projectId.ToString());
+            if (project == null)
+            {
+                return NuGetContractsFactory.CreateInstalledPackagesResult(InstalledPackageResultStatus.ProjectNotReady, packages: null);
+            }
+
+            var status = InstalledPackageResultStatus.Successful;
+            List<NuGetInstalledPackage> installedPackages;
+
+            if (project is BuildIntegratedNuGetProject packageReferenceProject)
+            {
+                var cacheContext = new DependencyGraphCacheContext();
+                var (packageSpecs, messages) = await packageReferenceProject.GetPackageSpecsAndAdditionalMessagesAsync(cacheContext);
+                if (messages?.Any(m => m.Level == LogLevel.Error) == true)
+                {
+                    status = InstalledPackageResultStatus.ProjectInvalid;
+                }
+
+                var packageSpec = packageSpecs.Single(s => s.RestoreMetadata.ProjectStyle == ProjectModel.ProjectStyle.PackageReference || s.RestoreMetadata.ProjectStyle == ProjectModel.ProjectStyle.ProjectJson);
+                var packagesPath = VSRestoreSettingsUtilities.GetPackagesPath(NullSettings.Instance, packageSpec);
+                FallbackPackagePathResolver pathResolver = new FallbackPackagePathResolver(packagesPath, VSRestoreSettingsUtilities.GetFallbackFolders(NullSettings.Instance, packageSpec));
+
+                var packageReferences = await project.GetInstalledPackagesAsync(cancellationToken);
+
+                installedPackages = packageReferences.Select(p => ToNuGetInstalledPackage(p, pathResolver))
+                    .ToList();
+            }
+            else if (project is MSBuildNuGetProject packagesConfigProject)
+            {
+                FolderNuGetProject packagesFolder = packagesConfigProject.FolderNuGetProject;
+
+                var packageReferences = await project.GetInstalledPackagesAsync(cancellationToken);
+
+                installedPackages = packageReferences.Select(p => ToNuGetInstalledPackage(p, packagesFolder))
+                    .ToList();
+            }
+            else
+            {
+                // unknown/unsupported project type
+                installedPackages = new List<NuGetInstalledPackage>(0);
+            }
+
+            return NuGetContractsFactory.CreateInstalledPackagesResult(status, installedPackages);
+        }
+
+        private NuGetInstalledPackage ToNuGetInstalledPackage(Packaging.PackageReference packageReference, FallbackPackagePathResolver pathResolver)
```

Comment:
What is a reason not to make this an extension method?

Reply:
What's a reason to use an extension method? :)   I don't know if it's still true, but when extension methods were first introduced, they could only be defined in static classes, which would mean that this method wouldn't be possible to keep in the same class as it's being used in. Given the `NuGetInstalledPackage` type is very specific to the `NuGetProjectService` contract, it doesn't seem to be a useful helper method for other classes to reuse.